text/0101-xcm-transact-allow-unlimited-weight.md

```diff
@@ -0,0 +1,95 @@
+# RFC-0101: XCM Transact shall also allow Unlimited weight
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 12 July 2024                                                                                |
+| **Description** | Enchance XCM Transact to allow `Unlimited` weight inner call                                |
+| **Authors**     | Adrian Catangiu                                                                             |
+
+## Summary
+
+The `Transact` XCM instruction currently forces the user to set a specific maximum weight allowed to the inner call and then also pay for that much weight regardless of how much the call actually needs in practice.
+
+This RFC proposes improving the usability of `Transact` by having the remote chain (which executes the `Transact`), get and charge the actual weight of the inner call from its dispatch info.
+
+## Motivation
+
+The UX of using `Transact` is poor because of having to guess/estimate the `require_weight_at_most` weight used by the inner call on the target.
+
+We've seen multiple `Transact` on-chain failures caused by guessing wrong values for this `require_weight_at_most` even though the rest of the XCM program would have worked.
+
+In practice, this parameter only adds UX overhead with no real practical value. Use cases fall in one of two categories:
+1. Unpaid execution of Transacts - in these cases the `require_weight_at_most` is not really useful, caller doesn't
+have to pay for it, and on the call site it either fits the block or not;
+2. Paid execution of _single_ Transact - the weight to be spent by the Transact is already covered by the `BuyExecution`
+weight limit parameter.
+
+We've had multiple OpenGov `root/whitelisted_caller` proposals initiated by core-devs completely or partially fail
+because of incorrect configuration of `require_weight_at_most` parameter. This is a strong indication that the
+instruction is hard to use.
+
+## Stakeholders
+
+- Runtime Users,
+- Runtime Devs,
+- Wallets,
+- dApps,
+
+## Explanation
+
+The proposed enhancement is simple: change `Transact` instruction:
+
+```diff
+- Transact { origin_kind: OriginKind, require_weight_at_most: Weight, call: DoubleEncoded<Call> },
++ Transact { origin_kind: OriginKind, weight_limit: WeightLimit, call: DoubleEncoded<Call> },
```

Comment:
Why do we need the `weight_limit` at all? 

I would just remove it?

Reply:
I am worried about potential use-cases where such an artificial limit might be required.

It's also potentially "breaking" if we directly remove it, as there is no conversion available, any existing program relying on that limit for some obscure reason will "break".

Both of the above are theoretical concerns, and in practice I expect nobody cares, but I would rather play it safe:
1. XCMv5: `Weight` -> `WeightLimit` - anyone who doesn't need artificial limit just uses `WeightLimit::Unlimited`, no hassle no worries
2. XCMv6+ : take a look at on-chain data, if we only `Unlimited`, we remove it

I am also not against removing it directly, looking for more feedback.

multi-step phase-out 👍  or kill it quick 👎 ?

Reply:
> It's also potentially "breaking" if we directly remove it, as there is no conversion available, any existing program relying on that limit for some obscure reason will "break".

The only reason I could see is to limit the amount of fees, but we probably already have some other instruction to limit the fees we are willing to pay?

Reply:
the proposed fee register should do exactly that. it shouldn’t be possible to charge more fee than then amount allocated into the fee register

Reply:
Yes, fees can be limited using dedicated fee mechanism, no need to mess around with weight limits for that.

I was thinking whether anyone needs to limit actual weight of "wild/unpredictable" Transact... but I can't think of a practical use-case, so I become more and more convinced we can go for it and directly remove it.

Reply:
yeah in the case we do want to limit the execution, the extrinsic should provide a mechanism anyway otherwise it will be unbenchable or have unreasonable weights. so we can basically assume such case should never happen

Reply:
done